### DISCUSSION TITLE
insert vs copy

### DISCUSSION THREAD

**Discussion by cxin-william** [sqlalchemy/sqlalchemy#7923] [Usage Questions]
Why is insert faster than copy when I use copy and INSERT?
300,000 pieces of data need_time：0:00:46.924713  copy
200,000 pieces of data need_time：0:00:49.978406  insert

2 million pieces of data need_time：0:09:34.036849 copy
2 million pieces of data need_time：0:08:26.001075 insert
`def write_to_table(df, schema, table_name, if_exists='append'):
    import io
    import pandas as pd
    from utils_all import create_connection
    db_engine = create_connection(**************************)
    string_data_io = io.StringIO()
    df.to_csv(string_data_io, sep='|', index=False)
    pd_sql_engine = pd.io.sql.pandasSQL_builder(db_engine)
    table = pd.io.sql.SQLTable(table_name, pd_sql_engine, frame=df,
                               index=False, if_exists=if_exists, schema=f'{schema}')
    table.create()
    string_data_io.seek(0)
    string_data_io.readline()  # remove header
    with db_engine.connect() as connection:
        with connection.connection.cursor() as cursor:
            copy_cmd = f"COPY {schema}.%s FROM STDIN HEADER DELIMITER '|' CSV" % table_name
            cursor.copy_expert(copy_cmd, string_data_io)
        connection.connection.commit()`

**Comment by CaselIT** [comment]
Hi.

You seem to be using psycopg2 directly, so you should probably ask directly on that project.
